### PR TITLE
[QtGui] Add option in gui to display Data Graph

### PR DESCRIPTION
**Pull request by epernod**
WIP project: Add window to display Data graph using [NodeEditor](https://github.com/paceholder/nodeeditor) library.

To use it, install NodeEditor.
check option in CMake: SOFA_BUILD_NODEGRAPH
and until better setup, manually set: 
- NODEEDITOR_INCLUDE_PATH to NodeEditor-install/include
- NODEEDITOR_LIBRARY_RELEASE  to  NodeEditor-install/lib/nodes.lib

DataGraph can be displayed in top menu: Edit->Data Graph...

graphData_04
graphData_03

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by epernod** [comment]
my bad, changes from PR #1181 are included here....

**Comment by epernod** [comment]
@damienmarchal for you, the last changes allow to retrieve pointer to BaseData (child and parent) which have been manually connected in the graph.
Ya plus qu'a dans runSofa2... comme on dit!

graphData_05
graphData_06

File: applications/sofa/gui/qt/CMakeLists.txt [hunk 1]
```diff
@@ -224,6 +224,32 @@ else()
...
+            dataGraph/SofaComponentNodeModel.h
+            )
+
+        list(APPEND SOURCE_FILES 
+            SofaWindowDataGraph.cpp
+            dataGraph/SofaComponentNodeModel.cpp
+            )
+    endif()
+endif()
+
```
Comment (guparan) [inline applications/sofa/gui/qt/CMakeLists.txt]: Find dependency packages before source files (l.65)
```cmake
# NodeEditor
option(SOFAGUIQT_ENABLE_NODEGRAPH "Build the Node graph representation. NodeEditor library is needed." OFF)
if(SOFAGUIQT_ENABLE_NODEGRAPH)
    find_package(NodeEditor REQUIRED)
endif()
sofa_set_01(SOFAGUIQT_HAVE_NODEEDITOR VALUE ${SOFAGUIQT_ENABLE_NODEGRAPH} BOTH_SCOPES)
```

Then add a if block in source files (here)
```cmake
# NodeEditor
if(SOFAGUIQT_ENABLE_NODEGRAPH)
    list(APPEND MOC_HEADER_FILES 
        SofaWindowDataGraph.h
        dataGraph/SofaComponentNodeModel.h
        )
    list(APPEND SOURCE_FILES 
        SofaWindowDataGraph.cpp
        dataGraph/SofaComponentNodeModel.cpp
        )
endif()
```

File: applications/sofa/gui/qt/CMakeLists.txt [hunk 2]
```diff
@@ -245,6 +272,10 @@ if(NOT SOFA_NO_OPENGL)
     endif()
 endif()
 
+if(NODEEDITOR_FOUND)
+    target_link_libraries(${PROJECT_NAME} PUBLIC ${NODEEDITOR_LIBRARY})
+endif()
+
```
Comment (guparan) [inline applications/sofa/gui/qt/CMakeLists.txt]:
```cmake
if(SOFAGUIQT_ENABLE_NODEGRAPH)
    target_link_libraries(${PROJECT_NAME} PUBLIC ${NODEEDITOR_LIBRARY})
    sofa_install_libraries(LIBRARIES ${NODEEDITOR_LIBRARY})
    target_include_directories(${PROJECT_NAME} PUBLIC "$<BUILD_INTERFACE:${NODEEDITOR_INCLUDES}>")
endif()
```

File: applications/sofa/gui/qt/RealGUI.cpp [hunk 3]
```diff
@@ -330,6 +333,9 @@ RealGUI::RealGUI ( const char* viewername)
       #if SOFAGUIQT_HAVE_QT5_CHARTS
       m_windowTimerProfiler(nullptr),
       #endif
+      #if SOFAGUIQT_HAVE_NODEEDITOR
+      m_sofaWindowDataGraph(nullptr),
+      #endif
```
Comment (guparan) [inline applications/sofa/gui/qt/RealGUI.cpp]: No indentation for preprocessor directives :-)

File: applications/sofa/gui/qt/CMakeLists.txt (hunk 1 above)
Comment (guparan) [reply applications/sofa/gui/qt/CMakeLists.txt]: What is tricky with `*_HAVE_*` variables is to keep them up-to-date when the option is disabled.
That is why `sofa_set_01` must stay out of the `if(SOFAGUIQT_ENABLE_NODEGRAPH)` block.

**Review by guparan — CHANGES_REQUESTED** [review]

File: applications/sofa/gui/qt/CMakeLists.txt [hunk 4]
```diff
@@ -245,6 +268,13 @@ if(NOT SOFA_NO_OPENGL)
     endif()
 endif()
 
+if(NODEEDITOR_FOUND)
+    include_directories(${NODEEDITOR_INCLUDE_PATH})
```
Comment (guparan) [inline applications/sofa/gui/qt/CMakeLists.txt]: This is redundant with the `target_include_directories` below (the latter is better).

File: applications/sofa/gui/qt/CMakeLists.txt [hunk 5]
```diff
@@ -245,6 +268,13 @@ if(NOT SOFA_NO_OPENGL)
     endif()
 endif()
 
+if(NODEEDITOR_FOUND)
+    include_directories(${NODEEDITOR_INCLUDE_PATH})     
+    target_link_libraries(${PROJECT_NAME} PUBLIC ${NODEEDITOR_LIBRARY})
+    sofa_install_libraries(LIBRARIES ${NODEEDITOR_LIBRARY})
+    target_include_directories(${PROJECT_NAME} PUBLIC "$<BUILD_INTERFACE:${NODEEDITOR_INCLUDES}>")
```
Comment (guparan) [inline applications/sofa/gui/qt/CMakeLists.txt]: So this should use NODEEDITOR_INCLUDE_PATH

**Review by guparan — CHANGES_REQUESTED** [review]

**Review by guparan — CHANGES_REQUESTED** [review]
Sorry @epernod but I didn't see that one last thing is missing: the *Config.cmake.in part of [my small tutorial](https://gist.github.com/guparan/0b0ed00ad60b740f85b794cdff255229).
With your PR, it should be done in SofaGuiConfig.cmake.in :wink:

**Comment by epernod** [comment]
@guparan here it is... Not 100% sure of what I did.
Is it:
```if(SOFAGUIQT_HAVE_NODEEDITOR)```  or ```if (SOFAGUIQT_ENABLE_NODEGRAPH)``` ?

just another question, in your tuto why is there ZLIB dependency SofaGeneral/SofaGeneralConfig.cmake.in and not in SofaGeneralLoader/SofaGeneralLoaderConfig.cmake.in ?

**Comment by guparan** [comment]
You should not have created a SofaGuiConfig.cmake.in in `sofa/gui/qt`. This file already exists and is in `<sofa_root>/SofaGui`. See my answer to your 2nd question below.

> @guparan here it is... Not 100% sure of what I did.
> Is it:
> `if(SOFAGUIQT_HAVE_NODEEDITOR)` or `if (SOFAGUIQT_ENABLE_NODEGRAPH)` ?

It's `if(SOFAGUIQT_HAVE_NODEEDITOR)` 

> just another question, in your tuto why is there ZLIB dependency SofaGeneral/SofaGeneralConfig.cmake.in and not in SofaGeneralLoader/SofaGeneralLoaderConfig.cmake.in ?

Because, from CMake point of view, SofaGeneral is a package and SofaGeneralLoader is not (it's only a target). 
A CMake package is something that can be loaded with `find_package` and that defines targets (most of the time those targets are libraries to link with `target_link_libraries`).
The XXXConfig.cmake.in file defines what the XXX package must do when someone calls `find_package(XXX)`. Typically, it will `find_package` the packages needed by it's own targets. Otherwise, at configure step, CMake will see that a target's dependency cannot be found for this package. See #1198 as example.

In SOFA, the packages are: 
  - the 8 main packages: SofaFramework, SofaSimulation, SofaBase, SofaCommon, SofaGeneral, SofaAdvanced, SofaMisc, SofaGui.
  - the plugins
  - the "pluginized" modules (targets that have become packages)

**Review by guparan — APPROVED** [review]

**Comment by guparan** [comment]
[ci-build][with-all-tests]

**Comment by epernod** [comment]
-> Ready? 🐱

**Pull request merged**